### PR TITLE
feat(optics): new alter lens to add or remove properties

### PR DESCRIPTION
**Pull request by serras**
Following the advice we got from internet, I've implemented an `alter` optic which works like the `at` in Haskell, that is, a lens to a `Maybe` which can add or remove properties. To make it composable I actually make it work over `Maybe s`, where `Maybe` is simply adding `notFound` as a possible input.

```haskell
alter : Key -> Lens (Maybe s) (Maybe a)
```

So now you have that:

```js
const nameL = optic(alter('flip'), alter('mix'))
expect(set(nameL, 'flop', {})).toEqual({ flip: { mix: 'flop' } })
```

We can always make `alter` the actual default to get the behavior we see in Ramda. Of course, that means that properties and indices in an array no longer work the same, as for the latter we cannot play this trick.

### CODE REVIEW

**Comment by serras** [comment]
I'll raise the coverage once you tell me whether: (1) you think this solves the problem (I do!), and (2) you think this should be the default instead of the "strict" lens, and (3) if so, whether you think `strict` is a good name for the other version of the lens which does not handling setting.

File: src/Lens.js [hunk 1]
```diff
@@ -62,3 +63,18 @@ export const prop = (key) => lens(get(key), set(key))
...
 // ix : Number → Lens s a
 export const ix = (index) => lens(get(index), set(index))
+
+// alter : String → Lens (Maybe s) (Maybe a)
+export const alter = (key) =>
+  lens(
+    (obj) => (isNotFound(obj) ? notFound : obj[key] || notFound),
+    (val, obj) => {
+      if (isNotFound(val)) {
+        var newObj = { ...obj }
```
Comment (kutyel) [inline src/Lens.js]: Never use `var`! 🚫

File: src/Lens.js [hunk 2]
```diff
@@ -62,3 +63,18 @@ export const prop = (key) => lens(get(key), set(key))
...
+        delete newObj[key]
```
Comment (kutyel) [inline src/Lens.js]: You can skip this `delete` by selectively destructuring the props you want to keep for the new object! 😉

**Review by kutyel — CHANGES_REQUESTED** [review]
Looks good, yes I think this is the way to go and that the other lenses `maybe` included should be explicit, this one being implicit when you construct an `optic`. This is what at least former Ramda lens users would expect from JS! 🤷‍♂️

File: src/Lens.js (hunk 2 above)
Comment (serras) [reply src/Lens.js]: I don't know how to do that. Could you give me a pointer/hint? :)

**Comment by serras** [comment]
It turns out that our use of `set` for arrays was wrong! In Ramda they use different functions for both, now I finally understand why (although my implementation is way cooler :P).

File: src/Lens.js (hunk 1 above)
Comment (serras) [reply src/Lens.js]: Fixed in 8fe59ab

File: src/Lens.js (hunk 2 above)
Comment (serras) [reply src/Lens.js]: Fixed in 8fe59ab

**Comment by serras** [comment]
By the way, I've renamed the `prop` lens to `mustBePresent` to make it more clear the behavior of that one.

This is not the shiniest PR, because I also had to add some more tests to make Coveralls happy.

**Review by kutyel — APPROVED** [review]
:shipit:

**Pull request merged**